### PR TITLE
feat: support envd run in local docker context

### PR DESCRIPTION
**Pull request by VoVAllen**
Signed-off-by: Jinjing.Zhou <allenzhou@tensorchord.ai>

### CODE REVIEW

**Comment by gaocegege** [comment]
Is it ready to review?

File: pkg/envd/docker.go [hunk 1]
```diff
@@ -187,6 +188,35 @@ func (e dockerEngine) ListImageDependency(ctx context.Context, image string) (*t
 	return dep, nil
 }
 
+func (e dockerEngine) getVerFromImageLabel(ctx context.Context, env string) (version.Getter, error) {
+	ctr, err := e.GetImage(ctx, env)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to inspect container: %s", env)
```
Comment (kemingy) [inline pkg/envd/docker.go:194]:
```suggestion
		return nil, errors.Wrapf(err, "failed to inspect image: %s", env)
```

File: pkg/envd/docker.go [hunk 2]
```diff
@@ -187,6 +188,35 @@ func (e dockerEngine) ListImageDependency(ctx context.Context, image string) (*t
...
+	if !ok {
+		return version.NewByVersion("v0"), nil
+	}
+	return version.NewByVersion(ver), nil
+}
+
+func (e dockerEngine) listEnvGeneralGraph(ctx context.Context, env string, g ir.Graph) (ir.Graph, error) {
+	ctr, err := e.GetImage(ctx, env)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to inspect container: %s", env)
```
Comment (kemingy) [inline pkg/envd/docker.go:206]:
```suggestion
		return nil, errors.Wrapf(err, "failed to inspect image: %s", env)
```

File: pkg/lang/ir/v0/compile.go [hunk 3]
```diff
@@ -166,6 +170,11 @@ func (g generalGraph) Labels() (map[string]string, error) {
 		return labels, err
 	}
 	labels[types.RuntimeGraphCode] = code
+	code, err = g.Dump()
+	if err != nil {
+		return labels, err
+	}
+	labels[types.GeneralGraphCode] = code
```
Comment (kemingy) [inline pkg/lang/ir/v0/compile.go:177]: I'm not sure about this design. Do we need to dump the whole graph? I thought we already have all the necessary information in labels.

File: pkg/types/label.go [hunk 4]
```diff
@@ -32,7 +32,9 @@ const (
 	ImageLabelCUDNN         = "ai.tensorchord.envd.gpu.cudnn"
 	ImageLabelContext       = "ai.tensorchord.envd.build.context"
 	ImageLabelCacheHash     = "ai.tensorchord.envd.build.digest"
+	ImageLabelSyntaxVer     = "ai.tensorchord.envd.syntax.version"
 	RuntimeGraphCode        = "ai.tensorchord.envd.runtimeGraph"
+	GeneralGraphCode        = "ai.tensorchord.envd.generalGraph"
```
[the remark this answers is not in the document]
Comment (VoVAllen) [reply pkg/types/label.go]: good idea. changed to your suggestion

File: pkg/envd/docker.go [hunk 5]
```diff
@@ -346,12 +376,29 @@ func (e dockerEngine) StartEnvd(ctx context.Context, so StartOptions) (*StartRes
 	base := fileutil.EnvdHomeDir(filepath.Base(so.BuildContext))
 	config.WorkingDir = base
 
+	var g ir.Graph
+	if so.Image == "" {
+		g = so.DockerSource.Graph
+	} else {
+		getter, err := e.getVerFromImageLabel(ctx, so.Image)
```
Comment (gaocegege) [inline pkg/envd/docker.go:383]: Could you please add some comments here?

**Review by gaocegege — APPROVED** [review]
LGTM :+1:
/lgtm

File: pkg/lang/ir/v0/compile.go (hunk 3 above)
Comment (kemingy) [reply pkg/lang/ir/v0/compile.go:177]: LGTM. We can create a new issue to track this.

**Review by kemingy — APPROVED** [review]

**Pull request merged**